cdap-app-fabric/src/main/java/io/cdap/cdap/internal/app/preview/PreviewDataSubscriberService.java

```diff
@@ -0,0 +1,261 @@
+  }
+
+  @Override
+  protected TransactionRunner getTransactionRunner() {
+    return transactionRunner;
+  }
+
+  @Nullable
+  @Override
+  protected String loadMessageId(StructuredTableContext context) throws Exception {
+    AppMetadataStore appMetadataStore = AppMetadataStore.create(context);
+    return appMetadataStore.retrieveSubscriberState(getTopicId().getTopic(), "preview.writer");
+  }
+
+  @Override
+  protected void storeMessageId(StructuredTableContext context, String messageId) throws Exception {
+    AppMetadataStore appMetadataStore = AppMetadataStore.create(context);
+    appMetadataStore.persistSubscriberState(getTopicId().getTopic(), "preview.writer", messageId);
+  }
+
+  @Override
+  protected void processMessages(StructuredTableContext structuredTableContext,
+                                 Iterator<ImmutablePair<String, PreviewMessage>> messages) throws Exception {
+    Map<PreviewMessage.Type, PreviewMessageProcessor> processors = new HashMap<>();
+
+    // Loop over all fetched messages and process them with corresponding PreviewMessageProcessor
+    while (messages.hasNext()) {
+      ImmutablePair<String, PreviewMessage> next = messages.next();
+      String messageId = next.getFirst();
+      PreviewMessage message = next.getSecond();
+
+      PreviewMessageProcessor processor = processors.computeIfAbsent(message.getType(), type -> {
+        switch (type) {
+          case DATA:
+            return new PreviewDataProcessor();
+          case STATUS:
+            return new PreviewStatusWriter();
+          case PROGRAM_RUN_ID:
+            return new PreviewProgramRunIdWriter();
+          default:
+            return null;
+        }
+      });
+
+      // noinspection ConstantConditions
+      if (processor == null) {
+        LOG.warn("Unsupported preview message type {}. Message ignored.", message.getType());
+        continue;
+      }
+      try {
+        processor.processMessage(message);
+        errorCount = 0;
+      } catch (Exception e) {
+        if (messageId.equals(erroredMessageId)) {
+          errorCount++;
+          if (errorCount >= maxRetriesOnError) {
+            LOG.warn("Skipping preview message {} after processing it has caused {} consecutive errors: {}",
+                     message, errorCount, e.getMessage());
+            continue;
+          }
+        } else {
+          erroredMessageId = messageId;
+          errorCount = 1;
+        }
+        throw e;
+      }
+    }
+  }
+
+  @Override
+  protected MessagingContext getMessagingContext() {
+    return messagingContext;
+  }
+
+  @Override
+  protected PreviewMessage decodeMessage(Message message) {
+    return GSON.fromJson(message.getPayloadAsString(), PreviewMessage.class);
+  }
+
+  /**
+   * The {@link PreviewMessageProcessor} for processing preview data.
+   */
+  private final class PreviewDataProcessor implements PreviewMessageProcessor {
+    PreviewDataProcessor() {}
```

Comment:
Remove this. You don't need to declare a no-op default constructor

Reply:
done.